artemis/reporting/base/report.py

```diff
@@ -0,0 +1,112 @@
+from artemis.utils import get_host_from_url, is_ip_address
+
+from .normal_form import NormalForm
+from .report_type import ReportType
+from .reporters import get_all_reporters
+
+
+@dataclass
+class Report:
+    # top_level_target is the target that was provided when adding targets to be scanned. It may not be the same as
+    # the target where actual vulnerability was found - e.g. you may start with scanning example.com and the
+    # vulnerability may be found on https://subdomain.example.com/phpmyadmin/
+    top_level_target: str
+
+    # The actual location of the vulnerability (it may be e.g. a domain or a URL)
+    target: str
+    report_type: ReportType
+
+    # Additional report data - the content depends on the report type
+    report_data: Dict[str, Any]
+    timestamp: Optional[datetime.datetime] = None
+    uuid: str = field(default_factory=lambda: str(uuid.uuid4()))
+
+    # IP address of the target - used later to deduplicate identical vulnerabilities on a domain and on an IP of the
+    # domain.
+    target_ip: Optional[str] = None
+
+    # Whether an attempt to resolve the IP of the target has been performed (the IP may be None because it has not been
+    # checked or because the domain didn't resolve).
+    target_ip_checked: bool = False
+
+    # Whether we already reported that vulnerability earlier
+    is_subsequent_reminder: bool = False
+
+    def __post_init__(self) -> None:
+        if not self.target_ip_checked:
+            # If something doesn't have :// it's a domain so let's skip obtaining IP as domain-related vulnerabilities
+            # (e.g. zone transfer or DMARC problems) don't have sensible IP versions.
+            if "://" in self.target:
+                host = get_host_from_url(self.target)
+                if is_ip_address(host):
+                    self.target_ip = host
+                else:
+                    try:
+                        self.target_ip = cached_gethostbyname(host)
+                    except socket.gaierror:
+                        self.target_ip = None
+            self.target_ip_checked = True
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Report):
+            return False
+
+        return (
+            self.top_level_target == other.top_level_target
+            and self.target == other.target
+            and self.report_type == other.report_type
+            and self.report_data == other.report_data
+        )
+
+    def target_is_ip_address(self) -> bool:
+        if "://" in self.target:
+            host = get_host_from_url(self.target)
+            return is_ip_address(host)
+        else:
+            return is_ip_address(self.target)
+
+    def alternative_with_ip_address(self) -> Optional["Report"]:
+        """If a report is about a URL where the host is a domain, not an IP, returns a version of this report
+        where the domain is replaced with an IP. Otherwise returns None."""
```

Comment:
Isn't this just misleading? I've seen plenty of misconfigurations where vulnerability exists only by IP, not by domain. We shouldn't do this.